analyser/module_analyser_call.c2

```diff
@@ -357,39 +402,48 @@ fn bool on_format_specifier(void* context, printf_utils.Specifier specifier, u32
         }
         return false;
     }
+    fa.out.add1(c);
 
+    fa.last_offset = offset + 1;
     fa.idx++;
     return true;
 }
 
-fn void Analyser.checkPrintArgs(Analyser* ma, Expr* format, u32 num_args, Expr** args, bool* change_format) {
-    // if format is StringLiteral or vardecl of type char[], it will wrapped in ArrayToPointerDecay
-    // otherwise it's a VarDecl of type (const) char* we cannot check and report this as an error
-    if (!format.isImplicitCast()) goto not_constant;
-    ImplicitCastExpr* ic = (ImplicitCastExpr*)format;
-    if (!ic.isArrayToPointerDecay()) goto not_constant;
-    format = ic.getInner();
-
-    SrcLoc format_loc;
+fn bool Analyser.checkPrintfArgs(Analyser* ma, Expr** format_ptr, u32 num_args, Expr** args) {
+    Expr* format = *format_ptr;
+    SrcLoc format_loc = format.getLoc();
     const char* format_text = printf_utils.get_format(format, &format_loc);
-    if (!format_text) goto not_constant;
+    if (!format_text) {
+        ma.error(format_loc, "format argument is not a constant string");
+        return false;
+    }
+
+    FormatAnalyser fa = { ma, format_text, args, num_args, 0, format_loc, 0 }
+    fa.out = string_buffer.create(256, false, 0);
```

Comment:
Put these 2 lines in FormatAnalyser.init()? removes all the constants here. Also add FormatAnalyser.free() to hide more internals. What do you think about adding a single FormatAnalyser in the ModuleAnalyser as member?

Reply:
I simplified the code.

I shall improve this part in a later patch, along with support for `scanf` format string analysis and conversion.